lib/internal/buffer.js

```diff
@@ -0,0 +1,86 @@
+'use strict';
+
+if (!process.binding('config').hasIntl) {
+  return;
+}
+
+const normalizeEncoding = require('internal/util').normalizeEncoding;
+const Buffer = require('buffer').Buffer;
+
+const icu = process.binding('icu');
+
+// Maps the supported transcoding conversions. The top key is the from_enc,
+// the child key is the to_enc. The value is the transcoding function.
+const conversions = {
+  'ascii': {
+    'latin1': (source) => {
+      return Buffer.from(source);
+    },
+    'utf8': (source) => {
+      return Buffer.from(source);
+    },
+    'utf16le': (source) => {
+      return icu.convertToUcs2('us-ascii', source);
+    }
+  },
+  'latin1': {
+    'ascii': (source) => {
+      return icu.convert('us-ascii', 'iso8859-1', source);
+    },
+    'utf8': (source) => {
+      return icu.convert('utf-8', 'iso8859-1', source);
+    },
+    'utf16le': (source) => {
+      return icu.convertToUcs2('iso8859-1', source);
+    }
+  },
+  'utf8': {
+    'ascii': (source) => {
+      return icu.convert('us-ascii', 'utf-8', source);
+    },
+    'latin1': (source) => {
+      return icu.convert('iso-8859-1', 'utf-8', source);
+    },
+    'utf16le': icu.convertToUcs2FromUtf8
+  },
+  'utf16le': {
+    'ascii': (source) => {
+      if (source.length % 2 !== 0)
+        throw new TypeError('Invalid UCS2 Buffer');
+      return icu.convertFromUcs2('us-ascii', source);
+    },
+    'latin1': (source) => {
+      if (source.length % 2 !== 0)
+        throw new TypeError('Invalid UCS2 Buffer');
+      return icu.convertFromUcs2('iso-8859-1', source);
+    },
+    'utf8': (source) => {
+      if (source.length % 2 !== 0)
+        throw new TypeError('Invalid UCS2 Buffer');
+      return icu.convertToUtf8FromUcs2(source);
+    }
+  }
+};
+
+// Transcodes the Buffer from one encoding to another, returning a new
+// Buffer instance.
+exports.transcode = function transcode(source, from_enc, to_enc) {
```

Comment:
Style: `s/from_enc/fromEncoding/` and `s/to_enc/toEncoding/`.  Ditto for cnv_from and cnv_to.